### DISCUSSION TITLE
Expanding the available Color Picker options from 32 to 54

### DISCUSSION THREAD

**Discussion by kyphii** [OpenRCT2/OpenRCT2#19428] [Feature requests] [answered]
The game's palette has 18 swatches for basic colors, each having 12 values. From these, the game can use 3 selections of the 12 values for normal, dark, and light variations*. Of the 54 theoretical variations, though, only 32 are usable in the color picker. I'm wondering if this could be expanded.

![image](https://user-images.githubusercontent.com/8711258/219790631-769c0f44-6e56-48e6-af16-c9a1e0c057cf.png)

Here you can see all the swatches and the colors in the game. The used ones are bordered in black, and the "unused" ones in gold (these are mockups). I think many of these colors look pretty appealing and would be welcome additions.

I've taken a cursory glance at the code and it seems like it might be possible to add these as new options in the color picker. It might've been limited to 32 in the original game for technical reasons, but now it seems like there's no such restriction (The colors are stored as a full byte rather than 5 bits, so should be no issue with park saves or TD6). 

The code seems complex and a bit all over the place, though, with many different enums and definitions across multiple files. Wondering if anyone more knowledgeable about the palette system has any input.

_\* Not all the swatches use the same mapping to generate these variations - There are at least 3 mappings each for the dark and light values, with black and white being totally unique. Light pink is also a notable exception to the whole system, as it uses a composite of colors from the bordeaux red and saturated red swatches - it's the only color like that. I've stuck it with the saturated reds in the diagram._

**Comment by duncanspumpkin** [comment]
There are ~3 sets of enums for colours. The 32 main colours, the extended colours (filter palette id) and finally the actual palette indexes. The extended colours are a super set of the main colours and handle a lot of the transparency related functions and contain a number of repeats of the main colours. Every extended colour has a palette map associated with it that is loaded from g1.dat. There are 144 extended colours and quite a few places assume it can fit in a byte. So if you want to add a new main colour you need to also add those respective parts. A lot of the codebase will likely expect the main colour repeats in the extended colours to be at certain offsets. I think that is where you would need focus on and make sure it's using the correct enum before trying to add more.

**Comment by kyphii** [comment]
Based on my (limited) understanding, I think the best approach for this concept would be to create a new palette definition file separate from g1, including both the original and new palettes. At the very least it would be the least bodgy solution. Maybe there'd be some issues with copyright though?

Currently I'm just messing around to get a feel for it. I've figured out how palettes are defined in g1 (it's kinda weird). Gonna keep experimenting and see if I can get the palette loading function referring to a new file in g2 instead of g1.

**Reply by duncanspumpkin** [reply]
> Based on my (limited) understanding, I think the best approach for this concept would be to create a new palette definition file separate from g1, including both the original and new palettes. At the very least it would be the least bodgy solution. Maybe there'd be some issues with copyright though?
> 
> Currently I'm just messing around to get a feel for it. I've figured out how palettes are defined in g1 (it's kinda weird). Gonna keep experimenting and see if I can get the palette loading function referring to a new file in g2 instead of g1.

That should be relatively easy enough to achieve. The palette maps don't need to be contiguous in g1 (and aren't contiguous in g1 in loco. Can't remember if rct2 is the same). So i don't see any reason to move it all into g2. Just add your new ones in g2.

**Comment by kyphii** [comment]
![image](https://user-images.githubusercontent.com/8711258/219907154-81cda146-8305-4ade-8f1d-56a95f396a78.png)

A closer look at all the colors in game. This is still a mockup (made by editing g1.dat directly). I might change some as there's a few that are quite similar. I'm also considering adding dedicated palettes for transparency and void, I think they would be useful for advanced players. They could be locked behind a cheat option.

**Reply by ocalhoun6** [reply]
Maybe "Invisible" is a better name than "transparent"? With 'transparent', I'd expect to see partial opacity, where you can see through it but still see it.

**Reply by dw1812** [reply]
A lot of those names aren't reflective of the colours described. "Beige" is peach to me, "khaki" is closer to a honeydew, and "lime green" is just a tad too fair to be called "lime green" - this is, of course, superficial however.

Would be nice to have these behind a toggle option too, for the sake of muscle memory and not overwhelming people not ready for the change.

**Reply by Gammadivided** [reply]
> Would be nice to have these behind a toggle option too, for the sake of muscle memory and not overwhelming people not ready for the change.

To expand on that, my suggestion would be to have 2 separate color dropdown options: one with a natural color progression (probably the default one), and another with the original color order first, followed by the new colors.

**Reply by duncanspumpkin** [reply]
> ![image](https://user-images.githubusercontent.com/8711258/219907154-81cda146-8305-4ade-8f1d-56a95f396a78.png)
> 
> A closer look at all the colors in game. This is still a mockup (made by editing g1.dat directly). I might change some as there's a few that are quite similar. I'm also considering adding dedicated palettes for transparency and void, I think they would be useful for advanced players. They could be locked behind a cheat option.

As i said transparency will have to be done at the same time. The game assumes that if you have a main colour palette map that you have a main colour transparency palette map. You can't have a transparent main colour that does not work. As for void you could have a flat colour palette map but it would work odd.

**Reply by kyphii** [reply]
> A lot of those names aren't reflective of the colours described. "Beige" is peach to me, "khaki" is closer to a honeydew, and "lime green" is just a tad too fair to be called "lime green" - this is, of course, superficial however.

These are all still working color names, so they're open to change. Maybe we'd also want to alter the existing color names to standardise them - Dark Green and Dark Water, for example, are actually the middle shades in their trios.

> To expand on that, my suggestion would be to have 2 separate color dropdown options: one with a natural color progression (probably the default one), and another with the original color order first, followed by the new colors.

There are some possibilities for this. See my next comment thread.

> As i said transparency will have to be done at the same time. The game assumes that if you have a main colour palette map that you have a main colour transparency palette map. You can't have a transparent main colour that does not work. As for void you could have a flat colour palette map but it would work odd.

Maybe I mislabeled it - the invisible paint scheme is just a normal palette with the primary remap colors all set to 0x00. I've been testing it in game by modifying g1.dat, and it works fine. It would have corresponding transparent and glass palettes. I haven't 100% figured out how those work yet, but the worst case is that those would just look weird (all the more reason to lock it behind a cheat option).

The use case for the invisible color is to hide parts of sprites, so it could be used to create decorations out of pieces of objects, easily hide supports (minus the footers), etc. The void color is maybe a bit more niche - custom blacktile palettes aren't quite as ubiquitous as they were prior to the void land texture, but they're still useful for some effects, so it'd be nice to have a dedicated paint scheme for it.

**Reply by duncanspumpkin** [reply]
> > A lot of those names aren't reflective of the colours described. "Beige" is peach to me, "khaki" is closer to a honeydew, and "lime green" is just a tad too fair to be called "lime green" - this is, of course, superficial however.
> 
> These are all still working color names, so they're open to change. Maybe we'd also want to alter the existing color names to standardise them - Dark Green and Dark Water, for example, are actually the middle shades in their trios.
> 
> > To expand on that, my suggestion would be to have 2 separate color dropdown options: one with a natural color progression (probably the default one), and another with the original color order first, followed by the new colors.
> 
> There are some possibilities for this. See my next comment thread.
> 
> > As i said transparency will have to be done at the same time. The game assumes that if you have a main colour palette map that you have a main colour transparency palette map. You can't have a transparent main colour that does not work. As for void you could have a flat colour palette map but it would work odd.
> 
> Maybe I mislabeled it - the invisible paint scheme is just a normal palette with the primary remap colors all set to 0x00. I've been testing it in game by modifying g1.dat, and it works fine. It would have corresponding transparent and glass palettes. I haven't 100% figured out how those work yet, but the worst case is that those would just look weird (all the more reason to lock it behind a cheat option).
> 
> The use case for the invisible color is to hide parts of sprites, so it could be used to create decorations out of pieces of objects, easily hide supports (minus the footers), etc. The void color is maybe a bit more niche - custom blacktile palettes aren't quite as ubiquitous as they were prior to the void land texture, but they're still useful for some effects, so it'd be nice to have a dedicated paint scheme for it.

That's going to behave very odd depending on how you do it i think it will behave differently for objects with 1 remap and objects with 2 remaps.

**Reply by kyphii** [reply]
![image](https://user-images.githubusercontent.com/8711258/219961353-2e07008d-e20d-4f3b-a191-8399bd090efe.png)
Seems to me like it behaves exactly as I'd expect - unless I'm missing what you're saying. What are you expecting would happen?

**Reply by ocalhoun6** [reply]
That screenshot also shows a couple great new uses for the invisible color -- that door without a wall around it could be very useful, as well as the tent top, which is now like a bunch of ribbons instead of a single striped piece of fabric. These are things that were previously impossible unless you used custom objects or maybe a hacked color palate. Excited to see what else is possible with this!

**Reply by duncanspumpkin** [reply]
> ![image](https://user-images.githubusercontent.com/8711258/219961353-2e07008d-e20d-4f3b-a191-8399bd090efe.png) Seems to me like it behaves exactly as I'd expect - unless I'm missing what you're saying. What are you expecting would happen?

I had a look into it. Turns out OpenRCT2 doesn't have the optimisations that we've added to OpenLoco so was getting confused with that. This wouldn't work with the OpenLoco version of this code.

**Reply by kyphii** [reply]
Ok. Are there any plans to add those/similar optimizations to OpenRCT2?

**Comment by 733737** [comment]
how would this change effect https://github.com/OpenRCT2/OpenRCT2/pull/16538

**Reply by duncanspumpkin** [reply]
Completely unrelated. This is about adding more palette maps. That is about palettes.

**Reply by 733737** [reply]
I apologize, I find this whole topic, having read over it thoroughly, quite confusing and overwhelming, Even though kyphii attempted to put it in layman's terms...

**Comment by kyphii** [comment]
![image](https://user-images.githubusercontent.com/8711258/219952414-71d84c02-76f7-45c2-af8d-afc800006277.png)

For the expanded color picker, there's plenty of possible configurations. If the cheat colors are inline with the rest, it could be 14x4 or 8x7. When they're disabled there would be empty cells for them, though, so I think it's better to have them on a separate line - in that case, the normal colors would be a 9x6 grid. We could either preserve the original color order and add the new colors around it, or group all the colors into swatches. Personally I like the latter option better, but I see the point about familiarity with the old layout. 

(Above are just examples - the order of colors shown here isn't set in stone either)

Of course, it could also be made into a new option to allow the player to choose their preference. With the new colors it'd definitely need a mapping table, since the order of the colors in memory no longer directly corresponds to their order in the picker. Might as well take advantage of that by having some different ordering options.

**Comment by 733737** [comment]
I am wondering whether more colors could be added on top of the 54 using the methodology involved in the creation of light pink

**Reply by kyphii** [reply]
The game has a hard limit of 256 palette maps (technically 251, as the water palette map take up multiple slots). My PR takes the number of used palette maps from 144 to 192. It's possible to make new colors by a composite of multiple color swatches - all that needs to be done is set the right indices in the primary remap range. There's a few reasons I didn't do this in the PR:

- Composite colors have odd results with custom palettes - you can already see this with light pink in certain palettes.
- Avoiding scope creep.
- Keeping it to 3 colors per swatch, I felt, retained the "feel" of RCT, still keeping a limited selection of colors.
- There's probably a limit on the number of composite colors that actually look good.

**Reply by 733737** [reply]
would it be prudent to reserve these swatches for the future, or should we fill them now? (or very soon in a 2nd pr) Maybe not you but someone else in the community would want to add more colors. We could put up a list of colors to a vote.

**Reply by spacek531** [reply]
What reason is there to?

**Reply by 733737** [reply]
FOMO? If we have the option to add even more colors, why would we settle for less?

**Reply by spacek531** [reply]
How is that FOMO?

**Reply by kyphii** [reply]
I don't think there needs to be any rush to expand the color set _even more_ - we've had the original set for 20 years, after all. I think the set of 54 is a good compromise for filling in the omissions from the original set, but not getting too crazy with it. Composite colors would probably be more niche, and would clutter up the color dropdown.

The PR does make it a fair bit simpler to add new palette maps in the future, but that isn't limited to new color options - they could be used for all sorts of visual effects.

**Reply by 733737** [reply]
maybe add a translucency color? This way we can make any wall into a window?

**Reply by kyphii** [reply]
Transparency is handled through special code and viewport flags - as far as I can tell that's not possible. At least, not without a ton more code changes, which I'm not too keen on at the moment.

**Reply by ocalhoun6** [reply]
I don't think translucency should be treated as a *color* per se. If you were going to attempt that -- attempt to make it possible for any recolorable object to be translucent -- it would be better to have the translucency as an entirely separate flag. That way, you could make "windows" tinted with any of the existing colors, just like the existing glass wall objects can do. (Though maybe the translucency options could be presented in the UI as if they're just additional colors.)

Honestly, though, that seems like too much work and too much headache. A simpler solution is probably to just make custom objects that are translucent. Heck, a lot of those probably already exist. That gives you complete freedom to properly integrate the quirks of each individual object, perhaps things like making some parts of them more or less transparent, depending on how thick the object is at that point -- or making some parts of the object translucent while keeping other parts opaque. So that seems like a solution that's both easier and better if you want translucent objects. (The exception would be if you want translucency for *ride* components, like translucent tracks or vehicles or supports. Though that could still be achieved with custom ride objects.)

**Reply by duncanspumpkin** [reply]
@kyphii 's invis is about as best as you can do its already relying on some quirks of the engine. It is not possible to do transparency that requires very different way of working. All objects with glass have two images one for the glass and one for the solid image  and the reason for this is to do with how colours and transparancy work in this engine.

**Comment by 733737** [comment]
TL;DR add 42 swatches locked behind cheats

I have a proposal I believe everyone will agree on. Add 42 swatches with no color assigned but lock them behind cheats menu. I see no major downsides to this.

Con: bunch of black squares on color picker when color cheat on

Pros: 1. ) Extended Custom Color palettes can be made
2.) does not change color picker for most users
3.) code doesn't have to be touched for years.
4.) colors or effects easier to add later in development if desired
5.) does not add any colors that might not be desirable
6.) compatible with all palettes
7.) purely optional to make a palette with either 54 or 55+ colors
8.) if devs add more official colors, custom palettes made before they were added will not be affected
9.) easy to implement, i think...

**Reply by spacek531** [reply]
What's the point of having 42 swatches that are all black?

**Reply by kyphii** [reply]
I don't see any reason to do this. Color options don't need to be 'reserved' in any way, and it's not possible for them to be user defined. The new colors are already compatible with custom palettes (they don't affect the structure of the palette itself), so no need to worry about palette makers having to update. In the future if even more color options were desired (I don't think it'd happen anytime soon), the changes made in the PR would make it fairly straightforward to add more.

**Reply by 733737** [reply]
they will only be black in the standard palette. They will be remappable for custom palettes so custom color 1-42 or remappable color 1-42

**Reply by kyphii** [reply]
That's not how it works, color options are hardcoded and are independent from the literal RGB values defined by the default/custom palette.

**Reply by spacek531** [reply]
> they will only be black in the standard palette. They will be remappable for custom palettes so custom color 1-42 or remappable color 1-42

The game has only 256 shades period. The colors you choose ingame are 12 specially-chosen shades of the 256. Brighter colors are made by choosing the brightest 6 shades and darker colors are made by choosing the darkest 6 shades. The 54 "extra" colors would have to use the existing 256 shades. If they were all black, it means they all use the same shade of the 256, and someone making a water object could change all 54 at the same time just by changing the one black shade they all use.

**Reply by 733737** [reply]
You're talking semantics. Can you add 42 swatches that have no color at all then?

**Reply by kyphii** [reply]
No, and why would we want to, and why 42?

**Reply by 733737** [reply]
I thought it would be nice to allow users of the game to have more colors available. I also thought 96 colors would line up in columns well in the color picker

**Comment by duncanspumpkin** [comment, marked as answer]
I'm closing this discussion now as it's now been converted to an issue.
